metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java

```diff
@@ -353,7 +353,7 @@ public ControllerResult<BrokerRegistrationReply> registerBroker(
         if (existing != null) {
             prevIncarnationId = existing.incarnationId();
             storedBrokerEpoch = existing.epoch();
-            if (heartbeatManager.hasValidSession(brokerId, existing.epoch())) {
+            if (heartbeatManager.hasValidSession(brokerId, existing.epoch()) && !existing.fenced() && !existing.inControlledShutdown()) {
                 if (!request.incarnationId().equals(prevIncarnationId)) {
                     throw new DuplicateBrokerRegistrationException("Another broker is " +
                         "registered with that broker id.");
```

Comment:
@cmccabe based on your offline comments - what if we keep the original behavior but just change this exception message to be more clear?

e.g.
```
"Another broker is registered with that broker id. This is benign if the broker was recently restarted, and will self-resolve shortly once the heartbeat manager expires the broker's session."
```

Reply:
@ahuang98 @cmccabe Note that the invalid session message and delay reregistration happens even after controlled shutdown. It is okay to have delays in reregistration during uncontrolled shutdown but not after controlled shutdown. Otherwise, this is a regression from the ZK implementation and behavior.